File: src/runtime/result_types.py
"""
Nexa Result Types — Error Propagation Infrastructure

v1.2: 实现 ? 操作符和 otherwise 内联错误处理的核心类型

设计参考: NTNT 的 ? 操作符和 otherwise 语句
- NexaResult: 类似 Rust Result<T, E>，成功(ok)或失败(err)
- NexaOption: 类似 Rust Option<T>，有值(some)或无值(none)
- ErrorPropagation: ? 操作符触发的内部异常，用于 early-return
- propagate_or_else: 统一处理 ? 和 otherwise 逻辑
- try_propagate: 在 flow 函数中捕获 ErrorPropagation 并转换为错误返回

关键设计原则:
1. Agent 优先: ? 和 otherwise 首先为 Agent.run() 结果设计
2. 向后兼容: 字符串结果自动包装为 NexaResult.ok
3. ErrorPropagation 是内部机制，不是用户可见的异常类型
4. 与 DAG ?? 条件分支操作符不冲突
"""


class ErrorPropagation(Exception):
    """? 操作符触发的错误传播异常，用于 early-return
    
    这是 Nexa 内部机制，不是用户可见的异常类型。
    当 ? 操作符遇到 NexaResult.err 或 NexaOption.none 时，
    抛出此异常，flow 函数的外层 try_propagate 会捕获并转换为错误返回。
    
    类似 Rust 的 ? 操作符工作机制：
    - Ok/Some → 继续执行（unwrap 返回值）
    - Err/None → 抛出 ErrorPropagation → flow 函数捕获 → 返回错误
    """
    def __init__(self, error):
        self.error = error
        super().__init__(str(error))


class NexaResult:
    """Nexa 结果包装器，类似 Rust 的 Result<T, E>
    
    表示一个可能成功或失败的操作结果：
    - NexaResult.ok(value) — 操作成功，包含返回值
    - NexaResult.err(error) — 操作失败，包含错误信息
    
    使用示例:
        result = NexaResult.ok("success data")
        result.is_ok  # True
        result.value   # "success data"
        
        result = NexaResult.err("something went wrong")
        result.is_err  # True
        result.error   # "something went wrong"
    
    ? 操作符核心:
        result.unwrap()  # 成功→返回值，失败→抛 ErrorPropagation
        
    otherwise 核心:
        result.unwrap_or("default")  # 成功→返回值，失败→返回默认值
    """
    
    def __init__(self, value=None, error=None, is_ok=True):
        self._value = value
        self._error = error
        self._is_ok = is_ok
    
    @staticmethod
    def ok(value):
        """创建成功结果"""
        return NexaResult(value=value, is_ok=True)
    
    @staticmethod
    def err(error):
        """创建失败结果"""
        return NexaResult(error=error, is_ok=False)
    
    @property
    def is_ok(self):
        """是否成功"""
        return self._is_ok
    
    @property
    def is_err(self):
        """是否失败"""
        return not self._is_ok
    
    @property
    def value(self):
        """成功时的值（仅在 is_ok 时有效）"""
        return self._value
    
    @property
    def error(self):
        """失败时的错误（仅在 is_err 时有效）"""
        return self._error
    
    def unwrap(self):
        """? 操作符的核心：成功返回值，失败则触发 ErrorPropagation
        
        这是 ? 操作符的工作原理：
        - NexaResult.ok → 返回内部值，继续执行
        - NexaResult.err → 抛出 ErrorPropagation，触发 early-return
        
        Returns:
            成功时的内部值
        
        Raises:
            ErrorPropagation: 失败时抛出，包含错误信息
        """
        if self._is_ok:
            return self._value
        raise ErrorPropagation(self._error)
    
    def unwrap_or(self, default):
        """otherwise 的核心：成功返回值，失败返回默认值
        
        这是 otherwise 的工作原理：
        - NexaResult.ok → 返回内部值
        - NexaResult.err → 返回 default
        
        Args:
            default: 失败时的默认值
        
        Returns:
            成功时的内部值，或失败时的默认值
        """
        if self._is_ok:
            return self._value
        return default
    
    def unwrap_or_else(self, handler):
        """otherwise 的函数版本：成功返回值，失败执行 handler
        
        Args:
            handler: 失败时的处理函数，接收 error 参数
        
        Returns:
            成功时的内部值，或 handler 的返回值
        """
        if self._is_ok:
            return self._value
        return handler(self._error)
    
    def map(self, fn):
        """映射成功值：成功时应用 fn，失败时保持不变
        
        Args:
            fn: 成功值的映射函数
        
        Returns:
            新的 NexaResult
        """
        if self._is_ok:
            return NexaResult.ok(fn(self._value))
        return self
    
    def map_err(self, fn):
        """映射错误值：失败时应用 fn，成功时保持不变
        
        Args:
            fn: 错误值的映射函数
        
        Returns:
            新的 NexaResult
        """
        if not self._is_ok:
            return NexaResult.err(fn(self._error))
        return self
    
    def and_then(self, fn):
        """链式操作：成功时应用 fn（fn 返回新的 NexaResult），失败时保持不变
        
        Args:
            fn: 成功值的链式函数，返回 NexaResult
        
        Returns:
            fn 返回的 NexaResult，或原始的 err
        """
        if self._is_ok:
            return fn(self._value)
        return self
    
    def or_else(self, fn):
        """链式错误恢复：失败时应用 fn（fn 返回新的 NexaResult），成功时保持不变
        
        Args:
            fn: 错误值的恢复函数，返回 NexaResult
        
        Returns:
            fn 返回的 NexaResult，或原始的 ok
        """
        if not self._is_ok:
            return fn(self._error)
        return self
    
    def __repr__(self):
        if self._is_ok:
            return f"NexaResult.ok({self._value!r})"
        return f"NexaResult.err({self._error!r})"
    
    def __str__(self):
        if self._is_ok:
            return str(self._value)
        return f"Error: {self._error}"
    
    def __eq__(self, other):
        if not isinstance(other, NexaResult):
            return False
        return self._is_ok == other._is_ok and (
            self._value == other._value if self._is_ok else self._error == other._error
        )


class NexaOption:
    """Nexa 可选值包装器，类似 Rust 的 Option<T>
    
    表示一个可能有值或无值的结果：
    - NexaOption.some(value) — 有值
    - NexaOption.none() — 无值
    
    使用示例:
        opt = NexaOption.some("hello")
        opt.is_some  # True
        opt.value    # "hello"
        
        opt = NexaOption.none()
        opt.is_none  # True
    """
    
    def __init__(self, value=None, is_some=True):
        self._value = value
        self._is_some = is_some
    
    @staticmethod
    def some(value):
        """创建有值的 Option"""
        return NexaOption(value=value, is_some=True)
    
    @staticmethod
    def none():
        """创建无值的 Option"""
        return NexaOption(is_some=False)
    
    @property
    def is_some(self):
        """是否有值"""
        return self._is_some
    
    @property
    def is_none(self):
        """是否无值"""
        return not self._is_some
    
    @property
    def value(self):
        """内部值（仅在 is_some 时有效）"""
        return self._value
    
    def unwrap(self):
        """? 操作符的核心：有值返回值，无值则触发 ErrorPropagation
        
        Returns:
            内部值
        
        Raises:
            ErrorPropagation: 无值时抛出
        """
        if self._is_some:
            return self._value
        raise ErrorPropagation(None)
    
    def unwrap_or(self, default):
        """otherwise 的核心：有值返回值，无值返回默认值
        
        Args:
            default: 无值时的默认值
        
        Returns:
            内部值或默认值
        """
        if self._is_some:
            return self._value
        return default
    
    def unwrap_or_else(self, handler):
        """otherwise 的函数版本：有值返回值，无值执行 handler
        
        Args:
            handler: 无值时的处理函数
        
        Returns:
            内部值或 handler 的返回值
        """
        if self._is_some:
            return self._value
        return handler()
    
    def map(self, fn):
        """映射有值情况：有值时应用 fn，无值时保持不变
        
        Args:
            fn: 值的映射函数
        
        Returns:
            新的 NexaOption
        """
        if self._is_some:
            return NexaOption.some(fn(self._value))
        return self
    
    def and_then(self, fn):
        """链式操作：有值时应用 fn（fn 返回 NexaOption），无值时保持不变
        
        Args:
            fn: 值的链式函数，返回 NexaOption
        
        Returns:
            fn 返回的 NexaOption，或原始的 none
        """
        if self._is_some:
            return fn(self._value)
        return self
    
    def or_else(self, fn):
        """链式无值恢复：无值时应用 fn，有值时保持不变
        
        Args:
            fn: 无值时的恢复函数，返回 NexaOption
        
        Returns:
            fn 返回的 NexaOption，或原始的 some
        """
        if not self._is_some:
            return fn()
        return self
    
    def to_result(self, error=None):
        """转换为 NexaResult：有值→ok，无值→err
        
        Args:
            error: 无值时的错误信息
        
        Returns:
            NexaResult
        """
        if self._is_some:
            return NexaResult.ok(self._value)
        return NexaResult.err(error)
    
    def __repr__(self):
        if self._is_some:
            return f"NexaOption.some({self._value!r})"
        return "NexaOption.none()"
    
    def __str__(self):
        if self._is_some:
            return str(self._value)
        return "None"
    
    def __eq__(self, other):
        if not isinstance(other, NexaOption):
            return False
        return self._is_some == other._is_some and (
            self._value == other._value if self._is_some else True
        )


# ============================================================
# 运行时辅助函数
# ============================================================

def propagate_or_else(result, otherwise_handler=None):
    """统一处理 ? 和 otherwise 逻辑
    
    这是 Nexa ? 操作符和 otherwise 语句的运行时核心函数。
    
    当只有 result 参数时（? 操作符模式）：
        - 成功 → 返回值
        - 失败 → 抛出 ErrorPropagation（early-return）
    
    当有 otherwise_handler 参数时（otherwise 模式）：
        - 成功 → 返回值
        - 失败 → 执行 otherwise_handler
    
    otherwise_handler 可以是:
        - 值（字符串、数字等）→ unwrap_or
        - NexaAgent 实例 → Agent.run() 作为 fallback
        - 函数/lambda → unwrap_or_else
        - dict（代码块）→ 执行代码块中的语句
    
    Args:
        result: NexaResult 或 NexaOption 实例
        otherwise_handler: otherwise 语句的处理器（可选）
    
    Returns:
        成功时的值，或 otherwise_handler 处理后的值
    
    Raises:
        ErrorPropagation: ? 操作符模式下失败时抛出
    """
    # 先检查是否已经是 NexaResult/NexaOption
    if isinstance(result, NexaResult):
        if result.is_ok:
            return result.value
        if otherwise_handler is not None:
            return _handle_otherwise(result.error, otherwise_handler)
        raise ErrorPropagation(result.error)
    
    if isinstance(result, NexaOption):
        if result.is_some:
            return result.value
        if otherwise_handler is not None:
            return _handle_otherwise(None, otherwise_handler)
        raise ErrorPropagation(None)
    
    # 向后兼容：非 NexaResult/NexaOption 值视为成功
    # Agent.run() 返回字符串时自动包装为 NexaResult.ok
    return result


def _handle_otherwise(error, handler):
    """处理 otherwise handler
    
    Args:
        error: 错误信息（来自 NexaResult.err 或 NexaOption.none）
        handler: otherwise 处理器
    
    Returns:
        handler 处理后的值
    """
    # handler 是 NexaAgent 实例 → 执行 Agent.run() 作为 fallback
    # 这是 Nexa 独有的特性：otherwise 可以指定另一个 Agent 作为 fallback
    from .agent import NexaAgent
    if isinstance(handler, NexaAgent):
        # Agent fallback: 调用 handler.run(str(error))
        # Agent.run() 本身会返回 NexaResult，所以需要 propagate_or_else
        agent_result = handler.run_result(str(error))
        if agent_result.is_ok:
            return agent_result.value
        # Agent fallback 也失败了 → 返回 Agent 的错误信息作为最终结果
        return str(agent_result.error)
    
    # handler 是函数/lambda → unwrap_or_else
    if callable(handler):
        return handler(error)
    
    # handler 是字典（代码块描述）→ 执行代码块中的语句
    if isinstance(handler, dict):
        # 代码块处理器 — 包含需要执行的语句列表
        # 这由 code_generator 生成的 Python 代码处理
        # 在运行时直接返回字典中的 "result" 字段
        if "result" in handler:
            return handler["result"]
        if "statements" in handler:
            # 执行语句列表（由生成的 Python 代码负责）
            return handler.get("fallback_value", str(error))
        return handler
    
    # handler 是值（字符串、数字等）→ 直接返回作为默认值
    return handler


def try_propagate(flow_func, *args, **kwargs):
    """在 flow 函数中捕获 ErrorPropagation 并转换为错误返回
    
    这是 flow 函数的外层包裹逻辑：
    - 正常执行 → 返回函数结果
    - ? 操作符触发 ErrorPropagation → 捕获并返回 NexaResult.err
    
    每个 flow 函数生成的 Python 代码都应该被 try_propagate 包裹，
    使得 ? 操作符的 early-return 行为正确工作。
    
    Args:
        flow_func: flow 函数
        *args: 函数参数
        **kwargs: 函数关键字参数
    
    Returns:
        函数正常返回值（可能是 NexaResult.ok 或原始值）
        或 NexaResult.err（当 ? 操作符触发 early-return）
    """
    try:
        result = flow_func(*args, **kwargs)
        # 如果结果已经是 NexaResult，直接返回
        if isinstance(result, NexaResult):
            return result
        # 否则包装为 NexaResult.ok
        return NexaResult.ok(result)
    except ErrorPropagation as e:
        # ? 操作符触发的 early-return → 转换为 NexaResult.err
        return NexaResult.err(e.error)
    except Exception as e:
        # 其他异常 → 也转换为 NexaResult.err（但不改变原有异常类型行为）
        # 注意：这里不捕获 ContractViolation 等用户可见的异常
        # 只有 ErrorPropagation 是内部机制
        from .contracts import ContractViolation
        from .type_system import TypeViolation
        if isinstance(e, (ContractViolation, TypeViolation)):
            raise  # 用户可见的异常，不转换
        return NexaResult.err(str(e))


def wrap_agent_result(raw_result):
    """将 Agent.run() 的原始返回值包装为 NexaResult
    
    向后兼容策略：
    - 如果已经是 NexaResult → 直接返回
    - 如果是字符串 → NexaResult.ok(string)
    - 如果是 Pydantic 模型 → NexaResult.ok(model)
    - 如果抛出异常 → NexaResult.err(exception_message)
    
    Args:
        raw_result: Agent.run() 的原始返回值
    
    Returns:
        NexaResult 包装
    """
    if isinstance(raw_result, NexaResult):
        return raw_result
    # 成功的字符串/Pydantic 模型结果 → NexaResult.ok
    return NexaResult.ok(raw_result)

File: avm/src/runtime/result_types.rs
//! Nexa Result Types — Error Propagation Infrastructure
//!
//! v1.2: 实现 ? 操作符和 otherwise 内联错误处理的核心类型
//!
//! 设计参考: NTNT 的 ? 操作符和 otherwise 语句
//! - NexaResult: 类似 Rust Result<T, E>，成功(ok)或失败(err)
//! - NexaOption: 类似 Rust Option<T>，有值(some)或无值(none)
//! - ErrorPropagation: ? 操作符触发的内部异常，用于 early-return
//!
//! 关键设计原则:
//! 1. Agent 优先: ? 和 otherwise 首先为 Agent.run() 结果设计
//! 2. 向后兼容: 字符串结果自动包装为 NexaResult::ok
//! 3. ErrorPropagation 是内部机制，不是用户可见的异常类型

use crate::vm::stack::Value;
use std::fmt;

/// ? 操作符触发的错误传播异常，用于 early-return
///
/// 这是 Nexa 内部机制，不是用户可见的异常类型。
/// 当 ? 操作符遇到 NexaResult::err 或 NexaOption::none 时，
/// 触发 ErrorPropagation，flow 函数的外层会捕获并转换为错误返回。
#[derive(Debug, Clone)]
pub struct ErrorPropagation {
    pub error: Value,
}

impl fmt::Display for ErrorPropagation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "ErrorPropagation: {}", self.error)
    }
}

/// Nexa 结果包装器，类似 Rust 的 Result<T, E>
///
/// 表示一个可能成功或失败的操作结果：
/// - NexaResult::ok(value) — 操作成功，包含返回值
/// - NexaResult::err(error) — 操作失败，包含错误信息
#[derive(Debug, Clone)]
pub enum NexaResult {
    Ok(Value),
    Err(Value),
}

impl NexaResult {
    /// 创建成功结果
    pub fn ok(value: Value) -> Self {
        NexaResult::Ok(value)
    }

    /// 创建失败结果
    pub fn err(error: Value) -> Self {
        NexaResult::Err(error)
    }

    /// 是否成功
    pub fn is_ok(&self) -> bool {
        matches!(self, NexaResult::Ok(_))
    }

    /// 是否失败
    pub fn is_err(&self) -> bool {
        matches!(self, NexaResult::Err(_))
    }

    /// ? 操作符的核心：成功返回值，失败则触发 ErrorPropagation
    ///
    /// 类似 Rust 的 ? 操作符工作机制：
    /// - Ok → 继续执行（返回内部值）
    /// - Err → 触发 ErrorPropagation → flow 函数捕获 → 返回错误
    pub fn unwrap(&self) -> Result<Value, ErrorPropagation> {
        match self {
            NexaResult::Ok(value) => Ok(value.clone()),
            NexaResult::Err(error) => Err(ErrorPropagation { error: error.clone() }),
        }
    }

    /// otherwise 的核心：成功返回值，失败返回默认值
    pub fn unwrap_or(&self, default: Value) -> Value {
        match self {
            NexaResult::Ok(value) => value.clone(),
            NexaResult::Err(_) => default,
        }
    }

    /// otherwise 的函数版本：成功返回值，失败执行 handler
    pub fn unwrap_or_else<F>(&self, handler: F) -> Value
    where
        F: FnOnce(&Value) -> Value,
    {
        match self {
            NexaResult::Ok(value) => value.clone(),
            NexaResult::Err(error) => handler(error),
        }
    }

    /// 映射成功值
    pub fn map<F>(&self, f: F) -> NexaResult
    where
        F: FnOnce(&Value) -> Value,
    {
        match self {
            NexaResult::Ok(value) => NexaResult::Ok(f(value)),
            NexaResult::Err(error) => NexaResult::Err(error.clone()),
        }
    }

    /// 映射错误值
    pub fn map_err<F>(&self, f: F) -> NexaResult
    where
        F: FnOnce(&Value) -> Value,
    {
        match self {
            NexaResult::Ok(value) => NexaResult::Ok(value.clone()),
            NexaResult::Err(error) => NexaResult::Err(f(error)),
        }
    }
}

impl fmt::Display for NexaResult {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            NexaResult::Ok(value) => write!(f, "NexaResult::Ok({})", value),
            NexaResult::Err(error) => write!(f, "NexaResult::Err({})", error),
        }
    }
}

/// Nexa 可选值包装器，类似 Rust 的 Option<T>
///
/// 表示一个可能有值或无值的结果：
/// - NexaOption::some(value) — 有值
/// - NexaOption::none() — 无值
#[derive(Debug, Clone)]
pub enum NexaOption {
    Some(Value),
    None,
}

impl NexaOption {
    /// 创建有值的 Option
    pub fn some(value: Value) -> Self {
        NexaOption::Some(value)
    }

    /// 创建无值的 Option
    pub fn none() -> Self {
        NexaOption::None
    }

    /// 是否有值
    pub fn is_some(&self) -> bool {
        matches!(self, NexaOption::Some(_))
    }

    /// 是否无值
    pub fn is_none(&self) -> bool {
        matches!(self, NexaOption::None)
    }

    /// ? 操作符的核心：有值返回值，无值则触发 ErrorPropagation
    pub fn unwrap(&self) -> Result<Value, ErrorPropagation> {
        match self {
            NexaOption::Some(value) => Ok(value.clone()),
            NexaOption::None => Err(ErrorPropagation { error: Value::Null }),
        }
    }

    /// otherwise 的核心：有值返回值，无值返回默认值
    pub fn unwrap_or(&self, default: Value) -> Value {
        match self {
            NexaOption::Some(value) => value.clone(),
            NexaOption::None => default,
        }
    }

    /// 映射有值情况
    pub fn map<F>(&self, f: F) -> NexaOption
    where
        F: FnOnce(&Value) -> Value,
    {
        match self {
            NexaOption::Some(value) => NexaOption::Some(f(value)),
            NexaOption::None => NexaOption::None,
        }
    }

    /// 转换为 NexaResult：有值→ok，无值→err
    pub fn to_result(&self, error: Value) -> NexaResult {
        match self {
            NexaOption::Some(value) => NexaResult::Ok(value.clone()),
            NexaOption::None => NexaResult::Err(error),
        }
    }
}

impl fmt::Display for NexaOption {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            NexaOption::Some(value) => write!(f, "NexaOption::Some({})", value),
            NexaOption::None => write!(f, "NexaOption::None"),
        }
    }
}

/// 统一处理 ? 和 otherwise 逻辑
///
/// 当只有 result 参数时（? 操作符模式）：
/// - 成功 → 返回值
/// - 失败 → 返回 ErrorPropagation
///
/// 当有 otherwise_handler 参数时（otherwise 模式）：
/// - 成功 → 返回值
/// - 失败 → 执行 otherwise_handler
pub fn propagate_or_else(result: &NexaResult, otherwise_handler: Option<&OtherwiseHandlerCtx>) -> PropagationResult {
    if result.is_ok() {
        // 成功 → 返回值
        PropagationResult::Ok(match result {
            NexaResult::Ok(value) => value.clone(),
            NexaResult::Err(_) => Value::Null, // 不应到达这里
        })
    } else {
        // 失败 → 根据是否有 otherwise handler 决定行为
        match otherwise_handler {
            Some(handler) => {
                // otherwise 模式：执行 handler
                PropagationResult::Fallback(handler.handle(result))
            }
            None => {
                // ? 操作符模式：触发 ErrorPropagation
                PropagationResult::Propagate(ErrorPropagation {
                    error: match result {
                        NexaResult::Err(error) => error.clone(),
                        NexaResult::Ok(_) => Value::Null, // 不应到达这里
                    }
                })
            }
        }
    }
}

/// otherwise handler 上下文
///
/// 描述 otherwise 语句右侧的处理器类型
#[derive(Debug, Clone)]
pub enum OtherwiseHandlerCtx {
    /// Agent 调用作为 fallback
    AgentCall { agent_name: String },
    /// 值作为默认 fallback
    Value(Value),
    /// 变量引用作为 fallback
    Variable(String),
    /// 代码块作为 fallback（包含语句列表）
    Block,
}

impl OtherwiseHandlerCtx {
    /// 处理 otherwise handler
    pub fn handle(&self, result: &NexaResult) -> Value {
        match self {
            OtherwiseHandlerCtx::AgentCall { agent_name } => {
                // Agent fallback: 在解释器层面处理
                Value::String(format!("fallback_agent:{}", agent_name))
            }
            OtherwiseHandlerCtx::Value(value) => value.clone(),
            OtherwiseHandlerCtx::Variable(name) => {
                Value::String(format!("fallback_var:{}", name))
            }
            OtherwiseHandlerCtx::Block => {
                // 代码块 fallback: 在解释器层面处理
                match result {
                    NexaResult::Err(error) => error.clone(),
                    _ => Value::Null,
                }
            }
        }
    }
}

/// 错误传播结果
#[derive(Debug, Clone)]
pub enum PropagationResult {
    /// 成功 → 继续执行
    Ok(Value),
    /// otherwise → 使用 fallback 值继续
    Fallback(Value),
    /// ? → 触发 ErrorPropagation (early-return)
    Propagate(ErrorPropagation),
}

/// 将原始 Agent 执行结果包装为 NexaResult
///
/// 向后兼容策略：
/// - Value::String → NexaResult::Ok(string)
/// - 其他值 → NexaResult::Ok(value)
pub fn wrap_agent_result(raw_result: Value) -> NexaResult {
    NexaResult::Ok(raw_result)
}

/// 将 NexaOption 转换为 NexaResult
pub fn option_to_result(opt: &NexaOption, error: Value) -> NexaResult {
    opt.to_result(error)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_nexa_result_ok() {
        let result = NexaResult::ok(Value::String("success".to_string()));
        assert!(result.is_ok());
        assert!(!result.is_err());
        
        let unwrapped = result.unwrap();
        assert!(unwrapped.is_ok());
        assert_eq!(unwrapped.unwrap(), Value::String("success".to_string()));
    }

    #[test]
    fn test_nexa_result_err() {
        let result = NexaResult::err(Value::String("error".to_string()));
        assert!(result.is_err());
        assert!(!result.is_ok());
        
        let unwrapped = result.unwrap();
        assert!(unwrapped.is_err());
        let propagation = unwrapped.unwrap_err();
        assert_eq!(propagation.error, Value::String("error".to_string()));
    }

    #[test]
    fn test_nexa_result_unwrap_or() {
        let ok_result = NexaResult::ok(Value::String("success".to_string()));
        let err_result = NexaResult::err(Value::String("error".to_string()));
        
        assert_eq!(ok_result.unwrap_or(Value::String("default".to_string())), 
                   Value::String("success".to_string()));
        assert_eq!(err_result.unwrap_or(Value::String("default".to_string())), 
                   Value::String("default".to_string()));
    }

    #[test]
    fn test_nexa_option_some() {
        let opt = NexaOption::some(Value::String("value".to_string()));
        assert!(opt.is_some());
        assert!(!opt.is_none());
        
        let unwrapped = opt.unwrap();
        assert!(unwrapped.is_ok());
        assert_eq!(unwrapped.unwrap(), Value::String("value".to_string()));
    }

    #[test]
    fn test_nexa_option_none() {
        let opt = NexaOption::none();
        assert!(opt.is_none());
        assert!(!opt.is_some());
        
        let unwrapped = opt.unwrap();
        assert!(unwrapped.is_err());
    }

    #[test]
    fn test_propagate_or_else_ok() {
        let result = NexaResult::ok(Value::String("success".to_string()));
        let propagation = propagate_or_else(&result, None);
        assert!(matches!(propagation, PropagationResult::Ok(_)));
    }

    #[test]
    fn test_propagate_or_else_err_with_handler() {
        let result = NexaResult::err(Value::String("error".to_string()));
        let handler = OtherwiseHandlerCtx::Value(Value::String("fallback".to_string()));
        let propagation = propagate_or_else(&result, Some(&handler));
        assert!(matches!(propagation, PropagationResult::Fallback(Value::String(_))));
    }

    #[test]
    fn test_propagate_or_else_err_no_handler() {
        let result = NexaResult::err(Value::String("error".to_string()));
        let propagation = propagate_or_else(&result, None);
        assert!(matches!(propagation, PropagationResult::Propagate(_)));
    }

    #[test]
    fn test_wrap_agent_result() {
        let raw = Value::String("agent output".to_string());
        let result = wrap_agent_result(raw);
        assert!(result.is_ok());
    }
}